src/Utils/LogRanker.ts

```diff
@@ -0,0 +1,132 @@
+
+        return this.logScores
+    }
+
+    private score(knownPath: boolean, validEndAction: boolean, numberUnknownUtterances: number, numberLowConfidence: number): number {
+        let score =
+            (knownPath ? 0 : 1)
+            + (validEndAction ? 0 : 1)
+            + Math.min(1, numberUnknownUtterances * 0.1)
+            + Math.min(1, numberLowConfidence * 0.1)
+        
+        return +(score / 4).toFixed(4)
+    }
+
+    // Returns true if TrainDialog's last action is in the set of actionIds
+    private isLastActionInSet(trainDialog: CLM.TrainDialog, actionIds: string[]) {
+        const lastRound = trainDialog
+            .rounds[trainDialog.rounds.length - 1]
+        
+        const actionId = lastRound.scorerSteps[lastRound.scorerSteps.length - 1].labelAction
+        if (!actionId) {
+            return false
+        }
+        return actionIds.includes(actionId)
+    }
+
+    // Return number user utterance in TrainDialog that are not included in the known utterance list
+    private numberUnknownUtterances(trainDialog: CLM.TrainDialog, knownUtterances: string[]): number {
+        const utterances = this.getUtterances(trainDialog)
+        return utterances.reduce((acc, u) =>
+            acc + (knownUtterances.includes(u) ? 0 : 1), 0)
+    }
+
+    // Return number scorer steps in TrainDialog with low confidence
+    private numberLowConfidence(logDialog: CLM.LogDialog, knownUtterances: string[]): number {
+        let numLowConfidence = 0
+        logDialog.rounds.forEach(r =>
+            r.scorerSteps.forEach(ss => {
+                if (ss.predictionDetails.scoredActions.length >= 2) {
+                    const scoredActions = ss.predictionDetails.scoredActions.sort(sa => sa.score)
+                    const scoreDiff = scoredActions[0].score - scoredActions[1].score
+                    if (scoreDiff < 0.2) {
+                        numLowConfidence = numLowConfidence + 1
+                    }
+                }
+            })
+        )
+        return numLowConfidence
+    }
+
+    // Return list of all unique utterances in TrainDialogs
+    private getAllUtterances(trainDialogs: CLM.TrainDialog[]): string[] {
+        const utterances: string[] = []
+        trainDialogs.forEach(td =>
+            td.rounds.forEach(r =>
+                r.extractorStep.textVariations.forEach(t => utterances.push(this.clean(t.text)))
+            )
+        )
+        // Make unique
+        return [...new Set(utterances)]
+    }
+
+    // Return list of all utternaces in a TrainDialog
+    private getUtterances(trainDialog: CLM.TrainDialog): string[] {
+        const utterances: string[] = []
+        trainDialog.rounds.forEach(r =>
+            r.extractorStep.textVariations.forEach(t => utterances.push(this.clean(t.text)))
+        )
+        return utterances
+    }
+
+    // Remove punctuation and lowercase
+    private clean(text: string): string {
+        const regex = /[!"#$%&'ʼ()*+,-./:;<=>?@[\]^_`{|}~]/g;
```

Comment:
Comment seems wrong re: lowercase.  Are you trying to strip out any non-alphanumeric data ?  If so, this would work better and be Unicode-aware

`const regex = /[^\p{L}]/gu`

Reply:
Not sure what you mean.  The function does lowercase too.   Just stripping punctuation

Reply:
Ah - I read the comment as "Remove punctuation and also remove lowercase characters".

Regex suggestion still applies.

Reply:
Would also say the method should be named what it does. If it removes punctuation and lowercase, name is `removePunctuationAndLowercase` so people don't wonder what `clean` means.